src/core/server/http/lifecycle/auth.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import Boom from 'boom';
+import { Lifecycle, Request, ResponseToolkit } from 'hapi';
+import { SessionStorage, SessionStorageFactory } from '../session_storage';
+
+enum ResultType {
+  authenticated = 'authenticated',
+  redirected = 'redirected',
+  rejected = 'rejected',
+}
+
+/** @internal */
+class AuthResult {
+  public static authenticated(credentials: any) {
+    return new AuthResult(ResultType.authenticated, credentials);
+  }
+  public static redirected(url: string) {
+    return new AuthResult(ResultType.redirected, url);
+  }
+  public static rejected(error: Error, options: { statusCode?: number } = {}) {
+    return new AuthResult(ResultType.rejected, { error, statusCode: options.statusCode });
+  }
+  public static isValidResult(candidate: any) {
+    return candidate instanceof AuthResult;
+  }
+  constructor(private readonly type: ResultType, public readonly payload: any) {}
+  public isAuthenticated() {
+    return this.type === ResultType.authenticated;
+  }
+  public isRedirected() {
+    return this.type === ResultType.redirected;
+  }
+  public isRejected() {
+    return this.type === ResultType.rejected;
+  }
+}
+
+/**
+ * @public
+ * A tool set defining an outcome of Auth interceptor for incoming request.
+ */
+export interface AuthToolkit {
+  /** Authentication is successful with given credentials, allow request to pass through */
+  authenticated: (credentials: any) => AuthResult;
+  /** Authentication requires to interrupt request handling and redirect to a configured url */
+  redirected: (url: string) => AuthResult;
+  /** Authentication is unsuccessful, fail the request with specified error. */
+  rejected: (error: Error, options?: { statusCode?: number }) => AuthResult;
+}
+
+const toolkit: AuthToolkit = {
+  authenticated: AuthResult.authenticated,
+  redirected: AuthResult.redirected,
+  rejected: AuthResult.rejected,
+};
+
+/** @public */
+export type Authenticate<T> = (
+  request: Request,
+  sessionStorage: SessionStorage<T>,
+  t: AuthToolkit
+) => Promise<AuthResult>;
```

Comment:
> I think this interface should be similar to Router's RequestHandler interface. 

Security is a very specific plugin, it has access to headers, which shouldn't be exposed via `KibanaRequest`. I'm going to introduce another mechanism to control access to header in following PR
https://github.com/elastic/kibana/pull/34631#discussion_r273135519
The current PR is already big enough.

>  In addition, it may be worth mirroring the ResponseFactory interface instead of AuthToolkit.

I was struggling to introduce similar interface or not. it's still interceptor mechanism and may have completely different outcomes, but okay let's try.

Reply:
Separate issues to do in follow-ups:
- introduce public api to register a route https://github.com/elastic/kibana/pull/34631#discussion_r273350400
- restrict access to Hapi Request object in `AuthenticationHandler`. Discuss with Security team what extension points they need to implement authorization logic (access to authorization header, a resource tags, etc.)
https://github.com/elastic/kibana/pull/34631#discussion_r273135519

possible improvements to discuss:
- add an extended error message for several `registerAuth` calls https://github.com/elastic/kibana/pull/34631#discussion_r274893965
- unify testing environment https://github.com/elastic/kibana/pull/34631#discussion_r272631873
- Boom usage https://github.com/elastic/kibana/pull/34631#discussion_r272629351 
- decide if `same site` cookie option is configurable https://github.com/elastic/kibana/pull/34631#discussion_r274275999

Reply:
> I'm going to introduce another mechanism to control this header in following PR.

SGTM, I didn't see the prior comment. We could probably construct KibanaRequest with a filtered set of headers for regular route handlers and only expose the Authenication header to AuthenticationHandlers.